### ISSUE TITLE
Unable to change rate or unit for track distance

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#13560]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Load NewDot on staging on an account with a workspace. 
2. Click Settings. 
3. Click Workspaces. 
4. Click into your workspace. 
5. Click Reimburse expenses. 
6. Click the Unit dropdown menu, and select a different unit, while online. 
7. Confirm you receive error message and unit does not change. 

## Expected Result:
Unit change and saved, with no error message for being offline. 

## Actual Result:
"An unexpected error occurred, please try again. If the error persists, please reach out to
[concierge@expensify.com](mailto:concierge@expensify.com)
.Your changes couldn't be saved. The workspace was modified while you were offline, please try again."

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.38-5
**Reproducible in staging?:** y
**Reproducible in production?:** y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
![Unit Bug](https://user-images.githubusercontent.com/43996225/207416231-0c8dda45-0d8b-4793-a462-3a59dcafeac4.jpg)

https://user-images.githubusercontent.com/43996225/207416270-3f63e199-0a4c-4cf5-be28-154a2163b4bf.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @begadtarek
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1670943899681949

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c6ee850ce85c0d80</li>
        <li>Upwork Job ID: 1605311368204656640</li>
        <li>Last Price Increase: 2022-12-20</li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
Triaging: 

From this SO: https://stackoverflow.com/c/expensify/questions/14418
- [ ] **The "bug" is actually a bug**: 
- [ ] **The bug is not a duplicate report of an existing GH (close the GH and add any novel details to the original GH instead)**: 
- [ ] **The bug is reproducible, following the reproduction steps:**
- [ ] **If you're unable to reproduce the bug, add the _Needs reproduction_ label. Comment on the issue outlining the steps you took to try to reproduce the bug, your results and tag the issue reporter and the Applause QA member who created the issue. Ask them to clarify reproduction steps and/or to check the reproduction steps again. Close issue.**: 
- [ ] **The GH template is filled out as fully as possible -- this means the GH body and title are clear (ie. could an external contributor understand it and work on it?):** 
- [ ] **The GH was created by an Expensify employee or a QA tester:** 
- [ ] **If the bug is an OldDot issue, you should decide whether it is widespread enough to justify fixing or it is better to wait for reunification. If it's better to wait, close the GH & provide this justification:** 
- [ ] **If there's a link to Slack, check the discussion to see if we decided not to fix it:** 
- [ ] **Decide if the GH should be resolved by an External contributor or Internal engineer, add the appropriate label:**

**Comment by joekaufmanexpensify** [comment]
Reproduced on staging web: 

![2022-12-13_14-09-12 (1)](https://user-images.githubusercontent.com/54866469/207424072-184394b8-4d64-45af-9d2a-3074022f8e41.gif)

**Comment by pecanoro** [comment]
I am going to see if I can see the cause of this bug since it seems it's reproducible!

**Comment by joekaufmanexpensify** [comment]
Sounds good! LMK if you plan to take this one/whether it should be internal/external.

**Comment by pecanoro** [comment]
Found the logs [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%22779127f548f4e72c-SAN%22)+AND+timestamp:[2022-12-13T18:37:44.641Z+TO+2022-12-13T20:37:44.641Z]&index=logs_expensify-019973)! This is probably internal since the error is coming from the back-end. However, I am still researching.

**Comment by joekaufmanexpensify** [comment]
👍

**Comment by jasperhuangg** [comment]
Can help take a look over here, I wrote the original command so I can probably see what's changed

**Comment by jasperhuangg** [comment]
It seems removing [this change to our Pusher service](https://github.com/Expensify/Web-Expensify/pull/35749/files#diff-1bfbc47d02c6a1bc3bcb7558f25dbb92d113e18dbf441e35be678a21cb9afefeR140-R144) fixes the problem. 

Somehow it's erroring out when encrypting the custom units data–specifically it's failing on the following line:

```
$payloadSize = mb_strlen($pusherCrypto->encrypt_payload($channels[0], $eventData), '8bit');
```

cc @Beamanator

**Comment by Beamanator** [comment]
Looks like there's even a few more places we may need to update:
- updateWorkspaceCustomUnit: https://github.com/Expensify/Web-Expensify/blob/0641db2b9d64efd90167a8e4c58e6f9c48790291/lib/PolicyAPI.php#L3338-L3339
- updateWorkspaceCustomUnitRate: https://github.com/Expensify/Web-Expensify/blob/0641db2b9d64efd90167a8e4c58e6f9c48790291/lib/PolicyAPI.php#L3743-L3744
- And possibly somewhere in the `addMembersToWorkspace` function, still looking for that fix

**Comment by jasperhuangg** [comment]
> updateWorkspaceCustomUnit: https://github.com/Expensify/Web-Expensify/blob/0641db2b9d64efd90167a8e4c58e6f9c48790291/lib/PolicyAPI.php#L3338-L3339
> updateWorkspaceCustomUnitRate: https://github.com/Expensify/Web-Expensify/blob/0641db2b9d64efd90167a8e4c58e6f9c48790291/lib/PolicyAPI.php#L3743-L3744

yup, I already have local changes fixing those @Beamanator 

Do you wanna tackle them all in one big PR?

**Comment by jasperhuangg** [comment]
@Beamanator Should we open a separate issue that includes this? Or should we just fix them all separately and link them to this PR?

**Comment by jasperhuangg** [comment]
@Beamanator Sounds like a plan, I'll also include some type of log that should fire next time someone tries to pass an associative array.

**Comment by jasperhuangg** [comment]
@Beamanator sameee hahah I kept reaching for it–here, you can have these in the meantime 😆 

👍👍👍👍👍👍👍👍👍👍👍

**Comment by jasperhuangg** [comment]
Weird, looks like Github isn't aware that I linked this issue in my PR 🤔 

Regardless here's the PR fixing this: https://github.com/Expensify/App/issues/13560

**Comment by Beamanator** [comment]
Noice @jasperhuangg ! I'll review! Looks like you are trying to point to this PR by the way :D https://github.com/Expensify/Web-Expensify/pull/35826

**Comment by pecanoro** [comment]
Well, I had a branch as well to fix it, please don't take over 😄

**Comment by pecanoro** [comment]
@jasperhuangg Please, unless it is a deploy blocker or so, next time wait for the original assigner to give you confirmation that you can take over so people don't do the same work twice.

**Comment by jasperhuangg** [comment]
Ahh I'm so sorry @pecanoro totally didn't realize. I did a PR recently related to this stuff so I was concerned I caused the issue. I'll be sure not to jump in like this next time, thanks for the feedback.

**Comment by pecanoro** [comment]
No idea why the PR is not getting linked, but here it is, some small problems with psalm: https://github.com/Expensify/Web-Expensify/pull/35830

**Comment by joekaufmanexpensify** [comment]
PR is merged!

**Comment by joekaufmanexpensify** [comment]
PR is on web staging.

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
PR is deployed to production!

**Comment by joekaufmanexpensify** [comment]
We handled this PR internally, so there are no PR/PR review payments to make. However, there is a $250 reporting bonus to pay to @BegadTarek. The PR is now subject to a 7 day regression waiting period before we make any payments. 

Given this issue was in an internal repo, there is no BZ/payment checklist, so going to copy them over manually!

**Comment by joekaufmanexpensify** [comment]
The solution for this issue has been 🚀 deployed to production 🚀 and is now subject to a 7-day regression period 📆. Here is the list of pull requests that resolve this issue:

https://github.com/Expensify/Web-Expensify/pull/35830

If no regressions arise, payment will be issued on 2022-12-27. 🎊

After the hold period, please check if any of the following need payment for this issue, and if so check them off after paying:

- [ ]  External issue reporter
- [ ]  Contributor that fixed the issue
- [ ]  Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:

Merged PR within 3 business days of assignment - 50% bonus
Merged PR more than 9 business days after assignment - 50% penalty

**Comment by joekaufmanexpensify** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ]  [@pecanoro] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@pecanoro] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ]  [@pecanoro] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: 
- [ ] [@joekaufmanexpensify] A [regression test](https://expensify.testrail.io/index.php?/dashboard) has been added or updated so that the same bug will not reach production again. Link to the GH issue for creating the test here:

**Comment by joekaufmanexpensify** [comment]
I don't think we need a new regression test for this. There is an existing regression test that covers this exact case (shown below). My thinking is a PR merged at some point which caused this regression. 

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/54866469/208768183-cf7c3b3d-57a9-4ee0-bbe7-cdca75e84bca.png">

**Comment by joekaufmanexpensify** [comment]
Temporarily adding the external label to add this job to upwork (so we can pay the reporting bonus).

**Comment by joekaufmanexpensify** [comment]
@BegadTarek Mind confirming your upwork handle/applying for [this job](https://www.upwork.com/jobs/~01c6ee850ce85c0d80) so we can pay you the $250 reporting bonus?

**Comment by joekaufmanexpensify** [comment]
Changing prioritization so we can issue payment and then complete the rest of BZ checklist next week

**Comment by joekaufmanexpensify** [comment]
@BegadTarek offer for $250 sent via upwork!

**Comment by joekaufmanexpensify** [comment]
[Looking for BZ volunteer](https://expensify.slack.com/archives/C01SKUP7QR0/p1671650924291199) to handle this payment on 2022-12-27 as I'm going to be OOO next week.

**Comment by joekaufmanexpensify** [comment]
@mallenexpensify is taking over as BZ here while I'm OOO next week. Thanks Matt! A few notes: 

- BZ portion of BZ checklist is all set. There are a few remaining steps that @pecanoro needs to handle. 
- This was internal PR in Expensify web repo. So the only payment needed is to the bug reporter. I just sent them a contract today. 
- Because this was in the Expensify web repo, the issue automation did not work. So I added the hold for payment and changed to weekly manually. So the issue will not change back to daily automatically on the 27th, even though that's when we'll be all set to issue payment. 

Let me know if you have any questions!

**Comment by pecanoro** [comment]
Ohh I forgot to update this. This was the offending [PR](https://github.com/Expensify/Web-Expensify/pull/34587) but I am wondering why we didn't catch it for 3 months so I am thinking maybe something else changed in the meantime, but I could not find it.

**Comment by pecanoro** [comment]
@mallenexpensify Are we free to close?

**Comment by mallenexpensify** [comment]
I think so, I removed my assignment and added Joe back, since I didn't do anything (and I have no idea how 'credit' works out for BZ stuff)

**Comment by joekaufmanexpensify** [comment]
@mallenexpensify Did you pay the contributor who reported? No worries if not, I can do it otherwise!

**Comment by joekaufmanexpensify** [comment]
Actually, I was able to confirm we have not yet paid this out in upwork: 

<img width="879" alt="image" src="https://user-images.githubusercontent.com/54866469/210660264-b4f72805-292b-4c8b-8a11-a9a2dc01c197.png">

**Comment by joekaufmanexpensify** [comment]
@BegadTarek $250 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
Closed upwork posting!

**Comment by joekaufmanexpensify** [comment]
Bug is fixed, BZ checklist complete, and payment sent. This is all set!

**Comment by mallenexpensify** [comment]
oh man.. I was the opposite of 'help'. 
Sorry I missed ya @BegadTarek

**Comment by BegadTarek** [comment]
Hahaha no worries, I thought there was probably a backlog due to the holidays.. hope you both have a happy new year
